proposals/0000-extension-lifecycle-framework.md

```diff
@@ -0,0 +1,128 @@
+ 
+ * `Legacy` extensions are explicitly not recommended for new code, much like `Deprecated`, however, they are expected to be supported indefinitely. This may be to maintain an older language set, maintain a bridge to a newer feature, or otherwise keep backward compatibility. Expected legacy extensions include `CUSKs`, `DeepSubsumption`,  and `NPlusKPatterns`.
+
+Above we list expected categorizations of several extensions based on statements in the user's guide. We do not attempt to prescribe any particular categorizations as part of this proposal and simply provide a small number of expectations from the perspective of someone reading the existing documentation.
+
+The expected extension lifecycle includes the following transitions:
+ * `Experimental` -> `Mature`
+ * `Experimental` -> `Deprecated`
+ * `Mature` -> `Legacy`
+ * `Mature` -> `Deprecated`
+ * `Legacy` -> `Deprecated`
+ * `Deprecated` -> `Legacy`
+
+However, it also seems plausible that new knowledge might from time to time cause a mature extension to once again be considered experimental, e.g. in the face of soundness bugs or subtle interactions with other features. We also do not rule out any transition explicitly to allow for unforeseen circumstances. 
+
+For existing, or future, language sets such as `GHC2021` or `Haskell98`, it is expected that none of the contained extensions would be `Experimental`. However, this proposal does not seek to impose any particular policy on the inclusion of extensions into language sets - the developers and the steering committee are always in the best position to make a decision about a concrete extension and extension set.
+
+### 2.2. User Interface
+
+Haskell users have different tolerances for risks related to language change. Some derive great value from using the newest features, while others are conservative. GHC should serve the whole spectrum.
+
+There will be an additional set of warnings:
+ * `-WXDeprecated`: Issue a warning when a deprecated extension is used
+ * `-WXMature`: Issue a warning when a mature extension is used
+ * `-WXExperimental`: Issue a warning when an experimental extension is used 
+ * `-WXLegacy`: Issue a warning when a legacy extension is used.
+
+Each category will also support the usual`-Wno-` syntax, so `-Wno-XDeprecated` will turn off warnings for deprecated extensions. Additionally, each extension should have a configurable warning that can be individually enabled or disabled, allowing users and teams to inform GHC about their local policies and needs. For instance, `-WXDeprecated -Wno-XRank2Types` would warn about all deprecated extensions except `Rank2Types`.
```

Comment:
What exactly do the `-WXFoo` warnings do? Do they just generically warn about the extension, or do they specifically say something about it's status? e.g. does `-WXRank2Types` produce something like
```
WARNING: Use of deprecated extension Rank2Types
```
Is that still the case if I don't have `-WXDeprecated` and I just turn on `-WXRank2Types`?

The reason I ask is that if those flags provide generically worded warnings about the use of an extension, then this actually provides us with an additional feature: configurable sets of warnings for use of language extensions. This is useful in a team setting where you maybe want to decide that you don't want to use some extension. You can do this today with `hlint`, but this would provide a built-in way of doing it.

I think a message like this might straddle the gap nicely?
```
WARNING: Use of extension Rank2Types (Deprecated)
```
i.e. the category is given as additional information, not necessarily pointed to as the cause of the warning.

Reply:
This is a good point. Our intention when writing this was that they give a warning for using the extension (which could, of course, be converted into an error), but we didn't propose a specific wording. We did want teams to be able to use it in the way you're suggesting, with individual warnings for individual extensions in order to enforce team policies in CI.

Depending on what's reasonable to implement, I wonder whether the right thing to do is to say which flag actually enabled the extension in the warning message, and render it based on that. E.g.
```
Foo.hs:3:5: warning: [-WXRank2Types]
    Use of extension Rank2Types
```
```
Foo.hs:3:5: warning: [-WXDeprecated]
    The extension Rank2Types is deprecated
```
with the latter taking precedence:
```
Foo.hs:3:5: warning: [-WXDeprecated, -WXRank2Types]
    The extension Rank2Types is deprecated
```

This may be infeasible or too expensive though.

Reply:
I think the square brackets should be consistent with the current behaviour. Currently GHC always shows the specific warning flag, plus the group(s) to which it belongs if `-fshow-warning-groups` is enabled. This is helpful because it is immediately obvious how to disable a warning selectively, even if it was enabled by a group. So I'd rather see something like:
```hs
Foo.hs:3:5: warning: [-WXRank2Types (in -WXDeprecated)]
    Use of extension Rank2Types, which is deprecated.

    This extension is an obsolete synonym for RankNTypes.

    Suggested fix: enable RankNTypes instead of Rank2Types.
```

```hs
Foo.hs:3:5: warning: [-WXGADTs (in -WXStable)]
    Use of extension GADTs, which is stable.
```

Reply:
Good point - I'll update the text after lunch to match your examples here.

Reply:
Done